generated/_typeshed_mypy_protobuf.pyi.expected

```diff
@@ -0,0 +1,25 @@
+# @generated by mypy-protobuf.  Do not edit manually!
+from google.protobuf.descriptor import (
+    FieldDescriptor as google___protobuf___descriptor___FieldDescriptor,
+)
+
+from google.protobuf.message import (
+    Message as google___protobuf___message___Message,
+)
+
+from typing import (
+    Generic as typing___Generic,
+    TypeVar as typing___TypeVar,
+)
+
+
+_ContainerMessageT = typing___TypeVar('_ContainerMessageT', bound=google___protobuf___message___Message)
+_ExtenderMessageT = typing___TypeVar('_ExtenderMessageT', bound=google___protobuf___message___Message)
+
+class _ExtensionFieldDescriptor(google___protobuf___descriptor___FieldDescriptor, typing___Generic[_ContainerMessageT, _ExtenderMessageT]):
+    pass
+
+class _ExtensionDict(typing___Generic[_ContainerMessageT]):
+    def __getitem__(self, extension_handle: _ExtensionFieldDescriptor[_ContainerMessageT, _ExtenderMessageT]) -> _ExtenderMessageT: ...
+
+    def __setitem__(self, extension_handle: _ExtensionFieldDescriptor[_ContainerMessageT, _ExtenderMessageT], value: _ExtenderMessageT) -> None: ...
```

Comment:
> 1 - Why does this need to be autogenerated? Seems like it potentially should belong either directly in typeshed itself, or inside the mypy_protobuf library as something to be imported

Yes I agree the autogeneration is not the best solution.  I can think of 3 possibilities:

* Generate a file as part of the protoc plugin (i.e. what we have currently).  I changed it to use a template file instead of having python code to write the file on the fly.  The benefit of this is that we can ensure these stubs will be present whenever `mypy` is run and don't add any new package requirements.

* As part of the `mypy_protobuf` library.  This simplifies the generated code and still ensures that the same version of the base classes used to generate stubs is used to evaluate them (assuming the user doesn't change library versions in between).  However, I think it also means that `mypy_protobuf` would need to be present whenever `mypy` is run, which is not currently a requirement.

* As part of the `typeshed`.  This addresses the issue of needing `mypy_protobuf` present when running `mypy`.  However, it means that we'd require the new `mypy` version to have these stubs be present.  It would also pose difficulties if we want to change anything about this file in the future.

There may be a way to do either of the latter two options in a "soft" way; i.e. fallback to the minimal extension support already in `master` if the respective dependency is not present.  I don't have a ton of experience publishing to the `typeshed` to know if/how they support API changes across package versions.

> 2 - Can you explain what value this is adding over a Dict[ContainerMessage, ExtenderMessage]

Two reasons.  `Message.Extensions` is not a dictionary, and does not support the various other methods associated with one (e.g. you can't do `foo.Extensions.values()`).  Additionally, I don't think using a static typing would allow for the polymorphism we currently have (i.e. `_ExtensionDict` is a class parametrized by `_ContainerMessageT` and `__{gs}estitem__` is a method of that class parametrized by `_ExtenderMessageT`).  The actual source code is [here](https://github.com/protocolbuffers/protobuf/blob/e667bf6eaaa2fb1ba2987c6538df81f88500d030/python/google/protobuf/internal/extension_dict.py#L63).  I'm not sure why the typeshed implementation does not include the other methods; I added all the public ones here.

Reply:
Ah! We can totally add these stubs to typeshed - as long as the stubs match the source code.
We have lots of precedent doing that (much of the google/protobuf typeshed stubs were contributed by dropboxers / folks who work on mypy-protobuf)

Source: https://github.com/protocolbuffers/protobuf/blob/e667bf6eaaa2fb1ba2987c6538df81f88500d030/python/google/protobuf/internal/extension_dict.py#L63

Typeshed: https://github.com/python/typeshed/tree/master/third_party/2and3/google/protobuf/internal
https://github.com/python/typeshed/tree/master/third_party/2and3/google/protobuf/internal

You can also test your typeshed change + mypy-protobuf change within mypy-protobuf by doing
```
CUSTOM_TYPESHED_DIR=~/path/to/your/typeshed/fork ./run_test.sh
```

Though one gotcha - when developing you should run the typeshed tests first - because errors within typeshed get suppressed when you use CUSTOM_TYPESHED_DIR.

It does mean that we'll have to wait for another release of mypy before we launch the next version of mypy-protobuf, but I think that's ok! #140 is similarly waiting on a typeshed change (you can look at that as an example)

Reply:
Interesting... I didn't realize the typeshed stubs for protobuf came from this same group (thank you!).  Are there any further comments you have on the current code before moving `typeshed.pyi` to the official typeshed?